### PR TITLE
SEA: Edge data action updates for v0.2

### LINKED ISSUE #51: SEA Edge Compute: Algorithm Updates
Existing metadata structure for SEA edge data products is overly cumbersome. Restructuring and reducing it will allow for easier bulk post-processing of edge data products. Some changes should also be made to the data result.

#### Data/Processing
- [x] Re-organize processing such that the IIR filter is applied before the FFT is computed. This is important since the calibration uses the filter, and the calibration result is applied to the IQ samples before the FFT even when the filter is not.
- [x] If possible, implement more aggressive parallelization to improve the efficiency and speed of edge computations. 
- [x] Add mean/max channel power component (2x15 pts for 15 channels), in addition to existing components.
- [x] Clean up logger for data product and calibration actions by stripping non-sigan parameters from those attempted to set on sigan
#### Metadata
- [x] Restructure metadata to be more concise and simple, grouping most metadata by channel.
- [x] Add sensor readouts to metadata: temperatures, humidity, preselector door sensor, etc.

### PR DESCRIPTION
**Pull request by aromanielloNTIA**
Resolves #51 

- Updates preselector package to 3.0.0, which increases default timeout to 3 seconds, adds retry-on-failure behavior, and allows for reading digital inputs from the preselector. This mitigates an ongoing error mode which occurs intermittently when communicating with the preselector.
- Updates `scos-tekrsa` to 2.1.0, which implements more efficient IQ streaming and ensures that all calibration parameters are correct when scaling IQ data. This results in a drastic reduction in IQ capture time variance and generally reduces IQ capture overhead time. 
- Further parallelizes data product processing, reducing data product compute time per-channel to approximately 5 seconds.
- Add single-value channel power (mean and max) values to metadata (in addition to the existing time series power data)
- Streamline metadata, putting most information into `captures` for each channel
- Add an `annotation` to metadata containing diagnostic sensor readouts from the preselector relay, SPU relay, and the NUC itself.
- Metadata changes reduce metadata size by ~50% per capture
- Calibration: calculate IIR filter ENBW on action instantiation, rather than during every action run. The ENBW depends only on the config so this is possible and removes a redundant computation
- Various changes reduce action runtime from ~85 to ~75 seconds.
- Make a hacky modification to `sigmf_builder` to allow for the desired metadata format, by allowing any dict to be added to a `capture`
- Various changes to make the logs more readable and useful

This PR is ready, but keeping as a draft until NTIA/scos-tekrsa#30 and NTIA/scos-tekrsa#28 are merged so everything can be on nasctn-sea branches.

### CODE REVIEW

**Review by dboulware — CHANGES_REQUESTED** [review]
Looks good. I think we will need to revisit the metadata changes more down the road, but for now I think we can live with only minor changes to what you have. For now, I suggest changing the core:version back to correspond to the SIgMF version, adding the namespace extensions back in, and I would move the items you have as annotations into the global object. Also, see inline comments regarding stop time, print statements, and treatment of SPU X410 switch in test_required_components.

File: scos_actions/actions/acquire_sea_data_product.py [hunk 1]
```diff
@@ -232,66 +280,39 @@ def generate_data_product(
...
+    for dp in all_results:
+        data_product.extend(dp)
 
-    tic = perf_counter()
-    data_product.extend(get_apd_results(iqdata, params))
     toc = perf_counter()
-    print(f"Got APD result @ {params[FREQUENCY]} in {toc-tic:.2f} s")
-
-    print(f"Got all data product @ {params[FREQUENCY]} results in {toc-tic1:.2f} s")
+    print(
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/actions/acquire_sea_data_product.py]: That is correct- removed them in [110ea8b](https://github.com/NTIA/scos-actions/pull/58/commits/110ea8ba3e3bd0c528653df1542ee9f82cba40d4)

File: scos_actions/actions/acquire_sea_data_product.py [hunk 2]
```diff
@@ -232,66 +280,39 @@ def generate_data_product(
...
 
-    data_product.extend(get_fft_results(iqdata, params))
-    toc = perf_counter()
-    print(f"Got FFT result @ {params[FREQUENCY]} in {toc-tic1:.2f} s")
-
-    tic = perf_counter()
     iqdata = sosfilt(iir_sos, iqdata)
     toc = perf_counter()
-    print(f"Applied IIR filter to IQ data @ {params[FREQUENCY]} in {toc-tic:.2f} s")
+    print(f"IIR filtered IQ data @ {params[FREQUENCY] / 1e6} MHz in {toc-tic:.2f} s")
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/actions/acquire_sea_data_product.py]: That is correct- removed them in [110ea8b](https://github.com/NTIA/scos-actions/pull/58/commits/110ea8ba3e3bd0c528653df1542ee9f82cba40d4)

File: scos_actions/actions/acquire_sea_data_product.py [hunk 3]
```diff
@@ -232,66 +280,39 @@ def generate_data_product(
 ) -> np.ndarray:
     """Process IQ data and generate the SEA data product."""
     # Use print instead of logger.debug inside ray.remote function
-    print(f"Generating data product @ {params[FREQUENCY]}...")
-    tic1 = perf_counter()
+    print(f"Generating data product @ {params[FREQUENCY] / 1e6} MHz...")
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/actions/acquire_sea_data_product.py]: That is correct- removed them in [110ea8b](https://github.com/NTIA/scos-actions/pull/58/commits/110ea8ba3e3bd0c528653df1542ee9f82cba40d4)

File: scos_actions/actions/acquire_sea_data_product.py [hunk 4]
```diff
@@ -449,6 +469,110 @@ def capture_iq(self, params: dict) -> dict:
...
+        Power/Control Box Temperature: oneWireSensor 1, set units to C
+        RF Box Temperature: oneWireSensor 2, set units to C
+        Power/Control Box Humidity: oneWireSensor 3
+
+        :param n_samps: The total number of data samples recorded.
+        """
+        tic = perf_counter()
+        # Read SPU sensors
+        for _, switch in switches.items():
+            if switch.name == "SPU X410":
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/actions/acquire_sea_data_product.py:504]: Agreed that this is not a great solution in the long run, but I like your recommendation for the time being. I added to test_required_components, which now throws an error if there is no switch named "SPU X410" in [b6249fd](https://github.com/NTIA/scos-actions/pull/58/commits/b6249fdf4c0f559664888d90451c3fbef0384d60)

File: scos_actions/actions/acquire_sea_data_product.py [hunk 5]
```diff
@@ -362,45 +379,48 @@ def __call__(self, schedule_entry, task_id):
...
-            cap_meta[rec_id]["sample_count"] = len(dp)
-
             # Generate metadata for the capture
-            self.create_channel_metadata(
-                rec_id, iteration_params[rec_id], cap_meta[rec_id], dp_idx
+            cap_meta[i].update(
+                {
+                    "sample_start": last_data_len,
+                    "sample_count": len(dp),
+                    "max_ch_pwr": max_ch_pwr,
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/actions/acquire_sea_data_product.py:399]: Leaving this to be addressed in the future

File: scos_actions/metadata/sigmf_builder.py [hunk 6]
```diff
@@ -3,15 +3,7 @@
 from sigmf import SigMFFile
 
 GLOBAL_INFO = {
-    "core:version": "v0.0.2",
-    "core:extensions": {
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/metadata/sigmf_builder.py]: Brought these back in [2d14ed1](https://github.com/NTIA/scos-actions/pull/58/commits/2d14ed133c9bc65c24fa31e96856bd5e38c9139b)

File: scos_actions/metadata/sigmf_builder.py [hunk 7]
```diff
@@ -3,15 +3,7 @@
...
-        "ntia-algorithm": "v1.0.0",
-        "ntia-core": "v1.0.0",
-        "ntia-environment": "v1.0.0",
-        "ntia-location": "v1.0.0",
-        "ntia-scos": "v1.0.0",
-        "ntia-sensor": "v1.0.0",
-    },
+    "core:version": "sea-prototype-v0.2",
```
[the remark this answers is not in the document]
Comment (aromanielloNTIA) [reply scos_actions/metadata/sigmf_builder.py]: Brought back original version numbers and replaced this with a new field `"ntia-nasctn-sea": "v0.2"` in [2d14ed1](https://github.com/NTIA/scos-actions/pull/58/commits/2d14ed133c9bc65c24fa31e96856bd5e38c9139b)

**Review by dboulware — APPROVED** [review]

**Pull request merged**